llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp

```diff
@@ -207,9 +207,31 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
   return true;
 }
 
+bool isLaneMask(Register Reg, MachineRegisterInfo *MRI,
+                const SIRegisterInfo &TRI) {
+  if (MRI->getType(Reg) != LLT::scalar(1))
+    return false;
+  const TargetRegisterClass *RC = MRI->getRegClassOrNull(Reg);
+  return RC && TRI.isSGPRClass(RC);
+}
+
 bool AMDGPUInstructionSelector::selectPHI(MachineInstr &I) const {
   const Register DefReg = I.getOperand(0).getReg();
   const LLT DefTy = MRI->getType(DefReg);
+  // Lane mask PHIs, PHI where all register operands have sgpr register class
+  // with S1 LLT, are already selected in divergence lowering pass.
+  if (I.getOpcode() == AMDGPU::PHI && isLaneMask(DefReg, MRI, TRI)) {
```

Comment:
Perhaps I should be clearer. I don't think we ever have a `PHI` here other than for lane masks. So it would be better for understanding the overall context to say something like:
```
  if (I.getOpcode() == AMDGPU::PHI) {
    assert(isLaneMask(DefReg, MRI, TRI);
    return true;
  }
```

Reply:
There is one regression test with PHI in `phi_s32_ss_sbranch - GlobalISel/inst-select-phi.mir `. It can be disabled. Other then that assert is fine.
I wanted to check isLaneMask for all operand, can I bring helper function back then? it would look like this without helper
```
  if (I.getOpcode() == AMDGPU::PHI) {
    assert(isLaneMask(DefReg, MRI, TRI));
    for (unsigned i = 1, e = I.getNumOperands(); i != e; e += 2)
      assert(isLaneMask(I.getOperand(i).getReg(), MRI, TRI));
    return true;
  }
```

Reply:
Huh, I wonder if @arsenm knows where that test comes from, whether that's on purpose or perhaps a mistake and the test should be removed.

If my understanding is wrong and other PHIs are possible, it would be good to understand why. If they aren't, then the test should be removed and it's not so important to check all the operands IMHO.

Reply:
Okay then, we can merge first two patches in the meantime